frontend/components/badge-examples.js

```diff
@@ -27,66 +30,70 @@ const ClickableCode = styled(StyledCode)`
-  static propTypes = {
-    definitions: PropTypes.arrayOf(serviceDefinitionPropType).isRequired,
-    baseUrl: PropTypes.string,
-    onClick: PropTypes.func.isRequired,
-  }
-
-  renderExample(exampleData) {
-    const { baseUrl, onClick } = this.props
-    const { title, example, preview } = exampleData
-
-    const { label, message, color, style, namedLogo } = preview
-    const previewUrl = staticBadgeUrl({
-      baseUrl,
-      label,
-      message,
-      color,
-      style,
-      namedLogo,
-    })
-
-    const { pattern, namedParams, queryParams } = example
-    const exampleUrl = badgeUrlFromPath({
-      baseUrl,
-      path: pattern,
-      namedParams,
-      queryParams,
-    })
+function Example({ baseUrl, onClick, exampleData }) {
+  const { title, example, preview } = exampleData
 
-    const key = `${title} ${previewUrl} ${exampleUrl}`
+  const { label, message, color, style, namedLogo } = preview
+  const previewUrl = staticBadgeUrl({
+    baseUrl,
+    label,
+    message,
+    color,
+    style,
+    namedLogo,
+  })
 
-    const handleClick = () => onClick(exampleData)
+  const { pattern, namedParams, queryParams } = example
+  const exampleUrl = badgeUrlFromPath({
+    baseUrl,
+    path: pattern,
+    namedParams,
+    queryParams,
+  })
 
-    return (
-      <tr key={key}>
-        <ClickableTh onClick={handleClick}>{title}:</ClickableTh>
-        <td>
-          <Badge clickable onClick={handleClick} src={previewUrl} />
-        </td>
-        <td>
-          <ClickableCode onClick={handleClick}>{exampleUrl}</ClickableCode>
-        </td>
-      </tr>
-    )
-  }
+  const handleClick = () => onClick(exampleData)
 
-  render() {
-    const { definitions } = this.props
+  return (
+    <tr>
+      <ClickableTh onClick={handleClick}>{title}:</ClickableTh>
+      <td>
+        <Badge clickable onClick={handleClick} src={previewUrl} />
+      </td>
+      <td>
+        <ClickableCode onClick={handleClick}>{exampleUrl}</ClickableCode>
+      </td>
+    </tr>
+  )
+}
+Example.propTypes = {
+  exampleData: examplePropType.isRequired,
+  baseUrl: PropTypes.string,
+  onClick: PropTypes.func.isRequired,
+}
 
-    const flattened = definitions.reduce((accum, current) => {
-      const { examples } = current
-      return accum.concat(examples)
-    }, [])
+export default function BadgeExamples({ definitions, baseUrl, onClick }) {
+  const flattened = definitions.reduce((accum, current) => {
+    const { examples } = current
+    return accum.concat(examples)
+  }, [])
 
-    return (
-      <ExampleTable>
-        <tbody>
-          {flattened.map(exampleData => this.renderExample(exampleData))}
-        </tbody>
-      </ExampleTable>
-    )
-  }
+  return (
+    <ExampleTable>
+      <tbody>
+        {flattened.map(exampleData => (
+          <Example
+            baseUrl={baseUrl}
+            exampleData={exampleData}
+            key={`${exampleData.title} ${exampleData.example.pattern}`}
```

Comment:
This all looks good, including in the review app. I'm not sure what's going on with a few things, like `key` for example. but that stems from my lack of familiarity with React.

Reply:
Identical adjacent elements need a unique key or React issues a warning. (It helps it reconcile the component tree with the DOM.) The key was generated before inside `renderExample` which was a helper function that returned a rendered component. The helper function has been replaced with a component, which means the key has to be moved here. Otherwise, what we'd have here would be a bunch of adjacent identical <Example> elements (regardless of what may be different inside them).